### PR TITLE
Device motion permission

### PR DESCRIPTION
**Pull request by KevinEverywhere**
**Description:**
iOS 13 introduced new permissions requirements for device motion and orientation. This PR enables iOS 13 Safari users to use device motion.

**Changes proposed:**
- A Scene Component addition (device-motion-permission-ui)
- CSS for screen visible selector to acquire permision
- Updated package.json file was required to build aframe

**Recommended Todo**
- Ensure that iPad 13 browsers also work (only Chrome does now)
- Always display this graphic for content that requires device motion

The second recommended todo is to proactively give user feedback that their device motion will be part of their experience. Depending on its function, this could have different opacity or color.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
     "ghpages": "ghpages -p gh-pages/",
     "lint": "semistandard -v | snazzy",
     "lint:fix": "semistandard --fix",
-    "precommit": "npm run lint",
+    "porecommit": "npm run lint",
```
Comment (dmarcos) [inline package.json]: Typo

File: package.json [hunk 2]
```diff
@@ -38,63 +38,63 @@
     "vendor/**/*"
   ],
   "dependencies": {
-    "custom-event-polyfill": "^1.0.6",
+    "custom-event-polyfill": "^1.0.7",
```
Comment (dmarcos) [inline package.json]: Better to not touch dependencies. Only upgrade when necessary.

File: src/components/index.js [hunk 3]
```diff
@@ -33,6 +33,7 @@ require('./windows-motion-controls');
 
 require('./scene/background');
 require('./scene/debug');
+require('./scene/device-motion-permission-ui');
```
Comment (dmarcos) [inline src/components/index.js]: Great. I think this is the good approach

File: src/components/scene/device-motion-permission-ui.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+/**
+ * UI for enabling device motion permission
+ */
+ module.exports.Component = registerComponent('device-motion-permission', {
+
+  dependencies: ['canvas'],
+
+  schema: {
+    enabled: {default: true},
+    enableFunc: {default:''}
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: What is this for?

File: src/components/scene/device-motion-permission-ui.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+var DEVICE_PERMISSION_CLASS = 'a-device-motion-permission';
+var DEVICE_PERMISSION_BTN_CLASS = 'a-device-motion-permission-button';
+
+/**
+ * UI for enabling device motion permission
+ */
+ module.exports.Component = registerComponent('device-motion-permission', {
+
+  dependencies: ['canvas'],
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: This dependency might not be necessary for this component

File: package.json (hunk 1 above)
Comment (KevinEverywhere) [reply package.json]: Ooops! That should have been removed. Before I updated dependencies in package.json, npm run lint was failing at several places. This has been reverted to "precommit".

File: package.json (hunk 2 above)
Comment (KevinEverywhere) [reply package.json]: I agree; and also, if package.json is updated, it will likely require moving forward the aframe version number. That said, if you use a clean machine and git clone the repo, you will find a number of issues with the current package.json file that makes updating pieces of it problematic. Can anyone else verify with a clean machine that package.json is stable? Thanks!

File: src/components/scene/device-motion-permission-ui.js (hunk 4 above)
Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: These are in flux and are not yet implemented. They may not belong in the final PR on this. I will revisit before resubmitting.

File: package.json (hunk 2 above)
Comment (dmarcos) [reply package.json]: Travis (our CI) machine does that per commit and haven't heard of any issues. Better to revert the package.json file to the original version and address any issues in a seperate PR

**Comment by dmarcos** [comment]
Thanks for doing this.. So far the code approach looks sensible. I haven't been able to test it on device yet.

File: src/components/scene/device-motion-permission-ui.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+var constants = require('../../constants/');
+
+var DEVICE_PERMISSION_CLASS = 'a-device-motion-permission';
+var DEVICE_PERMISSION_BTN_CLASS = 'a-device-motion-permission-button';
+
+/**
+ * UI for enabling device motion permission
+ */
+ module.exports.Component = registerComponent('device-motion-permission', {
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: It should be called `device-motion-permission-ui.js`

Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: Good catch--corrected.

Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: Good catch--corrected.

**Comment by KevinEverywhere** [comment]
Update on this: I found that iPad works for Firefox and Safari: in both cases, the applications had Desktop Version Requested. You either needed to set the preference in your preferences, or select the aA icon in the left of the location for Safari; or the three dots  on the right side of Firefox, to Request Mobile Site. That means that what I have done is working across the target browsers on iOS and iPad browsers; I believe that it will also work on Android devices as well (not tested on many). The controller is in the bottom left of the screen to ask permission. For android devices, you will not need to select it.

https://planetkevin.com/aframe/examples/showcase/device-motion-permission/a.html

**Comment by KevinEverywhere** [comment]
The last thing remaining from development perspective is getting past Travis. It is erring out on code that was written before me which I do not believe my code impacts. Is there something I am not understanding on this?

File: package.json (hunk 2 above)
Comment (KevinEverywhere) [reply package.json]: I've reverted the package.json and resolved all the other issues described. The big remaining question is, do we want to have a temporary or permanent visible icon on the bottom left, the way it is now; or, do we want to have a full screen interrupt, similar to 8thwall or vr-mode-ui when you are in portrait mode, to question device-motion-permission? I believe it is ready to be checked in.

Comment (dmarcos) [reply package.json]: Thanks for the hard work. I would say we want a modal dialog similar to 8thwall or vr-mode-ui. Button is easy to miss and hard to understand what it's for. We can apply and reuse vr-mode-ui  style for consistency.

**Comment by KevinEverywhere** [comment]
Matching Travis requirements, pulling out identified linting errors. Last big remaining item, making the modal screen attractive. Currently, Cancel and Continue buttons are functional. My idea: a Built with Aframe logo and wording that tells users that permission is required for device motion. I can continue with this, or if anyone else has an idea of what can be included here graphically, that could also be helpful.

**Comment by dmarcos** [comment]
Thanks for all the hard work! It's coming along pretty well. I made a few style changes to match linter style, `eslint-disable` should not be necessary. As mentioned the only remaining work is styling the dialog. I would keep it as unbranded as possible. We can reuse the rotation modal dialog gray background add cancel, accept buttons and some text: `This VR site requires access to your sensors. Press accept to allow`

Image from iOS

File: src/components/scene/device-motion-permission-ui.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+ *
+ * Structure: <div><button></div>
+ *
+ * @param {function} onClick - click event handler
+ * @returns {Element} Wrapper <div>.
+ */
+
+function createDeviceMotionPermissionWindow (onClick, obj) {
+  var wrapper, innerWrapper, aframeBuilt;
+  var cancelBtn, continueBtn;
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: We do one variable declaration per line and we avoid abbreviations: cancelButton and cotinueButton

File: src/components/scene/device-motion-permission-ui.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+var DEVICE_PERMISSION_CANCEL_CLASS = 'a-device-motion-permission-cancel';
+var BUILT_WITH_AFRAME_CLASS = 'a-built-with-aframe';
+
+/**
+ * UI for enabling device motion permission
+ */
+module.exports.Component = registerComponent('device-motion-permission-ui', {
+  schema: {
+    enabled: { default: true },
+    request: {default: {orientationChange:null, deviceOrientation:null}},
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: Does this need to be configurable via schema? Do we need to handle deviceOrientation events explicitly or we can just rely on CSS?

File: src/components/scene/device-motion-permission-ui.js [hunk 9]
```diff
@@ -1,4 +1,3 @@
-/* global DeviceOrientationEvent */
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: This is necessary for the linter

File: src/components/scene/device-motion-permission-ui.js (hunk 8 above)
Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: No, that was inadvertently left in from something earlier. What is now in place is the single component with the potential for one or both orientation events (device orientation and orientation change). The example on my site has two simple functions attached through the attributes, orientation-change and device-orientation. The drawback that I have on their usage at this point is that they use the window object to reference the functions. I'll hopefully get this done tomorrow with your edits. Thanks for your attention.

<a-scene device-motion-permission-ui orientation-change="noProof" device-orientation="proofOfWork">

https://planetkevin.com/aframe/examples/showcase/device-motion-permission/a.html

Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: `
function noProof(event){
--
  | console.log(event);
  | }
  | function proofOfWork(event){
  | console.log(event);
  | }
  | </script>
  |  
  | </head>
  | <body>
  | <a-scene device-motion-permission-ui orientation-change="noProof" device-orientation="proofOfWork">
  |  

`

**Comment by KevinEverywhere** [comment]
Just about complete. Has new modal screen; will add clickable to it later today (November 14th)

File: src/components/scene/device-motion-permission-ui.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    orientationChange: {default: null},
+    deviceOrientation: {default: null},
+    deviceMotionEl: { default: '' }
+  },
+  init: function () {
+    this.deviceMotionEl = null;
+    this.onDeviceMotionClick = bind(this.onDeviceMotionClick, this);
+    this.onOrientationChangeClick = bind(this.onOrientationChangeClick, this);
+    this.grantedDeviceMotion = bind(this.grantedDeviceMotion, this);
+    if (typeof window.orientation !== 'undefined') {
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: This if statement can be simplified to:
```
if (DeviceOrientationEvent && DeviceOrientationEvent.requestPermission) {
      this.deviceMotionEl = createDeviceMotionPermissionDialog(this.onDeviceMotionClick, this);
      this.el.appendChild(this.deviceMotionEl);
}
```
See [suggested edits](https://github.com/dmarcos/aframe/blob/b4be06e206821cedaa37a6b4512f4bd6641be1bd/src/components/scene/device-motion-permission-ui.js#L27)

File: src/components/scene/device-motion-permission-ui.js [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        'Your device and application combination do not support device motion events.'
+      );
+    }
+  },
+
+  grantedDeviceMotion: function () {
+    const orientationChange=this.el.getAttribute('orientation-change');
+    const deviceOrientation=this.el.getAttribute('device-orientation');
+    this.remove();
+      if(orientationChange){
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: We probably don't need handling orientation changes here

File: src/components/scene/device-motion-permission-ui.js [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * Create a button that when clicked will provide device motion permission.
+ *
+ * Structure: <div><button></div>
+ *
+ * @param {function} onClick - click event handler
+ * @returns {Element} Wrapper <div>.
+ */
+
+function createDeviceMotionPermissionWindow (onClick, obj) {
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: Rename to createDeviceMotionPermissionDialog, update argument names and docs above. [Suggested changes](https://github.com/dmarcos/aframe/blob/b4be06e206821cedaa37a6b4512f4bd6641be1bd/src/components/scene/device-motion-permission-ui.js#L69)

File: src/components/scene/device-motion-permission-ui.js (hunk 11 above)
Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: Agreed--before DeviceOrientationChange was included as a global, I was working around eslint errors. I'll include this in the next push.

File: src/components/scene/device-motion-permission-ui.js [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+  remove: function () {
+    if (this.deviceMotionEl) {
+      this.el.removeChild(this.deviceMotionEl);
+    }
+  },
+
+  /**
+   * Enable device motion permission when clicked.
+   */
+  onDeviceMotionClick: function () {
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: try catch not needed and redundant check of DeviceOrientationEvent. This method can be simplified. [Suggested edits](https://github.com/dmarcos/aframe/blob/b4be06e206821cedaa37a6b4512f4bd6641be1bd/src/components/scene/device-motion-permission-ui.js#L42)

File: src/style/aframe.css [hunk 14]
```diff
@@ -242,6 +242,96 @@ a-scene audio {
   display: block;
 }
 
+.a-device-motion-permission-full{
```
Comment (dmarcos) [inline src/style/aframe.css]: I would probably made it a [gray opaque background](https://user-images.githubusercontent.com/39342/68625009-fcb6ea80-048c-11ea-9142-bdf80987620c.png) covering the whole scene with text and two big buttons. Smaller dialog / window partially covering the scene is harder to read in particular in iframed or embedded mode.

**Comment by dmarcos** [comment]
Great job! Thanks for sticking with it. I made a few change suggestions. This is getting really close.

File: src/components/scene/device-motion-permission-ui.js [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    try {
+      /*eslint-disable */
```
Comment (dmarcos) [inline src/components/scene/device-motion-permission-ui.js]: Don't use `eslint-disable`. Linter issues should be addressed

File: src/components/scene/device-motion-permission-ui.js (hunk 10 above)
Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: The reasons for the structure here follow. I have tried these in different browsers and devices, including multiple iOS versions. There are other ways to do this, but going too simple or too complex will lead to similar outcomes. Thoughts?

iOS13
 - DeviceOrientationEvent.requestPermission needs to be called, not only tested to exist;
iOS and Android
 - grantedDeviceMotion is called by mobile devices (window.orientation)
Desktop Browsers
 - Do not call grantedDeviceMotion

Comment (dmarcos) [reply src/components/scene/device-motion-permission-ui.js]: DeviceOrientationEvent.requestPermission is already called when the user clicks on the allow button:
https://github.com/aframevr/aframe/pull/4303/files/8c30aaf2ce8b732dcc8b1644322fcad5f08e7036#diff-81b7296fca054a84c9266fd8fdebee05R75

Do we need to call it twice? 

lf think if `DeviceOrientationEvent.requestPermission` is not available we don't need to call any additional methods or initialize the dialog.

File: src/components/scene/device-motion-permission-ui.js (hunk 11 above)
Comment (dmarcos) [reply src/components/scene/device-motion-permission-ui.js]: Sorry, there must be something I'm not understanding. Why does orientation changes relate to the dialog? Is not using CSS sufficient? What are the attributes above for?

File: src/components/scene/device-motion-permission-ui.js (hunk 12 above)
Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: Done.

File: src/components/scene/device-motion-permission-ui.js (hunk 11 above)
Comment (dmarcos) [reply src/components/scene/device-motion-permission-ui.js]: What is the grantedDeviceMotion method for? It seems to be only handling orientation changes that are already handled by A-Frame.

**Comment by KevinEverywhere** [comment]
Just pushed latest updates; with comments and most all critiques included.

**Comment by subelsky** [comment]
thanks @KevinEverywhere! FYI I tested out your demo on Android 9 Chrome browser (& also on iOS 13 Safari), works well!

File: src/components/scene/device-motion-permission-ui.js (hunk 11 above)
Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: @dmarcos, grantedDeviceMotion unifies Android/iOS device orientation and orientation change methods. For a developer, it gives one entry point to ensure that event triggers are able to be handled in like fashion. It may be a moot issue for non-iOS 13 device/browsers combinations; it is a place where aframe events can be added confidently. The biggest use case is, if someone attempts to include Device Orientation behavior before permission is given, it might fail without the user knowing that it has failed. It provides a single place where events can be associated with user interaction; a place where aframe events may be attached with confidence.

File: src/style/aframe.css (hunk 14 above)
Comment (KevinEverywhere) [reply src/style/aframe.css]: Agreed; this is done in current commit.

File: src/components/scene/device-motion-permission-ui.js (hunk 15 above)
Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: This has been updated to omit eslint-disable from the code.

File: src/components/scene/device-motion-permission-ui.js (hunk 11 above)
Comment (dmarcos) [reply src/components/scene/device-motion-permission-ui.js]: Sorry, I'm still missing something. Could you help me understand by describing a concrete use case scenario? If there's an issue related to `orientationchange` events I feel it should be handled somewhere else. Regarding the solution, using global variable names passed through HTML attributes not associated to any component doesn't feel consistent with rest of A-Frame. I made a few cosmetic changes to the PR and removed for now. We can reintroduce if necessary.

**Comment by dmarcos** [comment]
Thanks a lot everybody for the hard work and the testing. I adjusted a bit the code style to the rest of A-Frame, removed some redundant logic and tweaked the visual style of the modal dialog. I tested on iOS 13.2 on iPhone X and Chrome m78 on a Pixel 3 with Android 10. I don't have an iPad handy at the moment to test with. If you can let me know if it works and what you think about the dialog. The PR is almost ready. We only need docs for the `device-motion-permission-ui` component and some basic test coverage.

dialog

File: src/components/scene/device-motion-permission-ui.js (hunk 11 above)
Comment (dmarcos) [reply src/components/scene/device-motion-permission-ui.js]: If the issue is notifying an application if DeviceOrientationEvent.requestPermission has succeeded or failed component events would be the way to go. e.g: `deviceorientationeventsgranted` / `deviceorientationeventsdenied`. Would that work?

**Comment by KevinEverywhere** [comment]
@dmarcos , will you want me to complete this: the readme, for instance. Also, I had one last test that is not included here. iPad OS 13 defaults to Request Desktop Website. A person with an iPad would not know that this fails without having feedback. my solution is below. I tested on a number of devices and operating systems, desktop and mobile. This took over 40 hours of time to resolve something that kept aframe content from being functional on iOS/iPad 13 devices. Thanks.

  isMobile: function() {
    return  /iPhone|iPad|iPod|Android/i.test(navigator.userAgent);
  },
  grantedDeviceMotion: function () {
    if (!this.isMobile()) {
      alert('You appear to have a mobile device, but you may have requested a Desktop Website. Check the location bar of your browser and ensure you are requesting the Mobile Website.');
    } else {
...
}

The other place this would be used is after the window.orientation check:

      // This is for desktop browsers and those who cancel or reject
      if( this.isMobile() ){
        alert('You appear to have a mobile device, but you may have requested a Desktop Website. Check the location bar of your browser and ensure you are requesting the Mobile Website.');
      } else {
        this.remove();
      }

File: src/components/scene/device-motion-permission-ui.js (hunk 11 above)
Comment (KevinEverywhere) [reply src/components/scene/device-motion-permission-ui.js]: Both orientation change and device orientation are granted permission by the same event, DeviceMotionEvent. That is why I included them both. Regarding the global variable name; I completely agree with you. I would think that there might be a way to pass the functions that depend on access to device motion to grantedDeviceMotion. That way, it is ensured that iOS/iPad will not have issues associated with attempting to include functionality before they have permission. One way or another, and it could be deviceorientationeventsgranted / deviceorientationeventsdenied, the developer should have access to a flag to know that it is possible to add a deviceorientation or orientationchange event. Thanks, and I am glad that I have been able to contribute to something that I needed to work! :)

**Comment by dmarcos** [comment]
I added an alert dialog recommending the user to enable mobile browsing if Safari for iPad is on desktop mode. If someone with an iPad + iOS 13 can verify that this is working it will be super appreciated.

**Comment by KevinEverywhere** [comment]
URL?

**Comment by dmarcos** [comment]
@KevinEverywhere can you fetch the PR and test it locally? 

`npm run start:https` and any of the examples will do. e.g: `https://10.1.10.146:9000/examples/boilerplate/hello-world/`

**Comment by KevinEverywhere** [comment]
I have an iPad and have tested this quite a bit. It appears that testing navigator.platform and multitouch will target iPad users whether or not they have Desktop Website Requested functioning. Also, it does not detect other tablet users. Window.orientation gives you that information. If you have a direct link to the built aframe.js file I could give it a test. In the meantime, this URL should work on Android and iPad tablets, as well as mobile phones of either platform.  This was submitted 12 days ago, and I have included the relevant JavaScript.  PS: I cannot access my dev network from my iPad but I can reference a URL of built aframe.js file.

https://planetkevin.com/aframe/examples/showcase/device-motion-permission/b.html

  isMobile: function() {
    return  /iPhone|iPad|iPod|Android/i.test(navigator.userAgent);
  },
init: function () {
    if (typeof window.orientation !== 'undefined') {
// Desktop and Desktop Website Requested (iPad, Android, as well as phones)
    } else {
      if( this.isMobile() ){   // This test is for mobile devices who do not have orientation enabled
        alert('You appear to have a mobile device, but you may have requested a Desktop Website. Check the location bar of your browser and ensure you are requesting the Mobile Website.');
      } else {
        this.remove();
      }


> Update on this: I found that iPad works for Firefox and Safari: in both cases, the applications had Desktop Version Requested. You either needed to set the preference in your preferences, or select the aA icon in the left of the location for Safari; or the three dots on the right side of Firefox, to Request Mobile Site. That means that what I have done is working across the target browsers on iOS and iPad browsers; I believe that it will also work on Android devices as well (not tested on many). The controller is in the bottom left of the screen to ask permission. For android devices, you will not need to select it.
> 
> https://planetkevin.com/aframe/examples/showcase/device-motion-permission/a.html

**Comment by dmarcos** [comment]
Adding iPad to `isMobile` function will result in tablets entering stereo cardboard mode vs. fullscreen that I don't think we want. Trying to find the simpler set of conditions that also preserve previous functionality. I'll get a hold of an iPad with iOS 13 later in the day and test the PR. Thanks for the help.

**Comment by subelsky** [comment]
@dmarcos am grabbing my iPad now and cloning the PR to test

**Comment by KevinEverywhere** [comment]
@subelsky and @dmarcos; if anyone can post the output file for testing, that would be great. If either you or anyone else can compare what already is up at the URL below to work on all target browser/devices; I think that the cardboard/fullscreen worry will be mooted. Also, you can test your iPhone against this. Currently, if your iPhone is set to Desktop Website Requested, you do not have feedback to let you know. What I have is reverse compatible and covers iOS/iPad and Android.

https://planetkevin.com/aframe/examples/showcase/device-motion-permission/b.html

**Comment by subelsky** [comment]
@KevinEverywhere here's what I did:
```
HOST=0.0.0.0 npm run dev
ngrok http 9000

**Comment by subelsky** [comment]
attached is the build file which I produced via: `npm run build`:
[aframe.js.gz](https://github.com/aframevr/aframe/files/3860791/aframe.js.gz)

hope that helps! this feature is going to be awesome for a couple of projects, PSYCHED

**Comment by dmarcos** [comment]
I changed the `mobile device requesting desktop site` condition to what @KevinEverywhere suggested and It works on an iPhone X with iOS 13 when requesting a desktop site. Also works on Android m78 / Pixel 3 / Android 10 when requesting mobile and desktop sites. All looks good. I'm going to write some tests and docs and I'll merge later in the day. We can adjust edge cases later if needed. Thanks all for your help.

**Comment by KevinEverywhere** [comment]
Looks great, @dmarcos; checked it on iPad Pro with iPad 13 installed with Mobile and Desktop Website Request and it worked as expected on both.

**Comment by dmarcos** [comment]
Documented and tested! Shipping 🚢 Thanks everybody for the hard work.

**Comment by Dirk-27** [comment]
@dmarcos : Is this feature coming with 0.9.3 on 6th december?

**Comment by dmarcos** [comment]
Yes but you can use it today via master builds: https://cdn.jsdelivr.net/gh/aframevr/aframe@934b8d358a2c458efdc8d65fcb6946f7303f0b51/dist/aframe-master.min.js

Let us know if you have any issues

**Comment by tommensink** [comment]
Works very well! Is there any chance this will be included in Aframe 0.8.2? Since the breaking change of <a-animation> in  V0.9.2 my code does not work in that version, and I see no real possiblity to upgrade it in the near future. I suspect a lot of people will stay with 0.8.2 for the coming months/years because of this breaking chance

**Comment by dmarcos** [comment]
@tommensink Unfortunately, We don’t patch previous releases. It’s not sustainable to maintain versions. What’s the breaking change that it’s preventing to upgrade to 0.9.2? a-animation got replaced by the [animation component](https://aframe.io/docs/0.9.0/components/animation.html#sidebar) but the API is the same. It should be easy to upgrade.

**Comment by tommensink** [comment]
Thanks, and I understand. It is just that I use <a-animation> a lot and would not know where to begin to replace it in my code, since I use a nested structure with some sort of animation sequencer that I made. You can see it in https://hh-webar.firebaseapp.com/Presentation_2019-01-26/. (Just let the head rotate on your screen to see the different transparancy sequences kicking in).

**Comment by tommensink** [comment]
My personal opinion is that moves from Apple like this motion-issue (which of course also can be understood, but nevertheless) and breaking Aframe chances will throw it back for some years to be adopted by the professional market. I for one will do the major serious AR/VR projects in Unity and use Aframe only for gadgets like dancing business cards

**Comment by KevinEverywhere** [comment]
As a developer, the choice between narrower proprietary (higher paying) and
open-source (open, agreed-upon standards) solutions is never-ending. I
disagree with the idea that this throws aframe back years. It merely drives
innovation and should keep us vigilant and on our toes. It will be the
community as much as the vendor that will keep technologies current.
Nothing against Unity or profit-motive companies. I just like the idea of a
large and common playing field.

**Comment by tommensink** [comment]
I hope I don't hijack this thread for this discussion too much, but one last thing then: You should hear me about breaking changes in Unity or their horrific tight integration of libraries like Vuforia (who break the API every release). But having said that, I am a full time AR developer, make dozens of apps as lead developer in four AR startups, including one of the first to do actual brain surgery with a HoloLens. So I have to choose the most appropriate and stable Entity Component Framework regularly. Funny moves like this one from Apple (and we had something similar by Android beginning of the year which requires the user to change a setting https://github.com/jeromeetienne/AR.js/issues/369#issuecomment-482047211) is one reason to choose another platform that will keep running no matter the OS change. An issue like this means that I will judge again in one or two years for Aframe in critical apps, hence my "Aframe is thrown back years" by this. And if the Aframe community decides they should make breaking changes to the API like <a-animation>, then this does not work in favour of this decision. Just my opinion, do think the major companies have similar considerations, but cannot speak for them. 
But I think you guys are awesome, this solution is great Kevin (I applied your first one which works with V0.8.2) and I always support AFrame in discussions and presentations as THE best open platform.
